### PR TITLE
PP-4352: Refactor out Capture logic ready for Stripe Integration

### PR DESCRIPTION
**Pull request by IreneLau-GDS**
## WHY
At present, all PaymentProvider classes implement PaymentProvider interface which constitutes signatures for multiple features (authorise, capture, refund, cancel and also some helper methods specific to feature). This is leading to `1. Fat interface 2. Multiple responsibilities for Payment Provider classes and 3. Having to work on same code base to deliver features (authorise, capture ..) in parallel`. 

As a first step, refactored `Capture` functionality

## WHAT
We introduced CaptureHandler interface and its concrete classes. As a result, we refactored all the capture logic into the relevant classes and the *PaymentProvider classes now only need to call
the capture hander to do the job.

This will help deliver functionality independently. For example, Capture feature for stripe can be implemented with new StripeCaptureHandler without need to change a lot in StripePaymentProvider

## HOW DID WE TEST THE CHANGES
Since we should not have changed any business logic because it is just pure refactoring, rerun all the exists test should be sufficient.

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/gateway/CaptureHandler.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package uk.gov.pay.connector.gateway;
+
+import uk.gov.pay.connector.gateway.model.request.CaptureGatewayRequest;
+import uk.gov.pay.connector.gateway.model.response.BaseCaptureResponse;
+import uk.gov.pay.connector.gateway.model.response.GatewayResponse;
+
+public interface CaptureHandler<T extends BaseCaptureResponse> {
+
+    public GatewayResponse<T> capture(CaptureGatewayRequest request);
```
Comment (oswaldquek) [inline src/main/java/uk/gov/pay/connector/gateway/CaptureHandler.java]: we don't need a `public` identifier here

File: src/test/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProviderTest.java [hunk 2]
```diff
@@ -75,7 +75,7 @@ public void shouldNotAuthoriseIfPaymentProviderReturnsNon200HttpStatusCode() {
     @Test
     public void shouldCapture() {
         mockPaymentProviderResponse(200, successCaptureResponse());
-        GatewayResponse<EpdqCaptureResponse> response = provider.capture(buildTestCaptureRequest());
+        GatewayResponse<EpdqCaptureResponse> response = epdqCaptureHandler.capture(buildTestCaptureRequest());
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProviderTest.java]: It feels like this test is no longer testing the EpdqPaymentProvider. what do you think about moving it to EpdqCaptureHandlerTest?

File: src/test/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProviderTest.java [hunk 3]
```diff
@@ -84,15 +84,15 @@ public void shouldCapture() {
     @Test
     public void shouldNotCaptureIfPaymentProviderReturnsUnexpectedStatusCode() {
         mockPaymentProviderResponse(200, errorCaptureResponse());
-        GatewayResponse<EpdqCaptureResponse> response = provider.capture(buildTestCaptureRequest());
+        GatewayResponse<EpdqCaptureResponse> response = epdqCaptureHandler.capture(buildTestCaptureRequest());
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProviderTest.java]: ditto previous comment

File: src/test/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProviderTest.java [hunk 4]
```diff
@@ -84,15 +84,15 @@ public void shouldCapture() {
...
         assertThat(response.isFailed(), is(true));
         assertThat(response.getGatewayError().isPresent(), is(true));
     }
 
     @Test
     public void shouldNotCaptureIfPaymentProviderReturnsNon200HttpStatusCode() {
         mockPaymentProviderResponse(400, errorCaptureResponse());
-        GatewayResponse<EpdqCaptureResponse> response = provider.capture(buildTestCaptureRequest());
+        GatewayResponse<EpdqCaptureResponse> response = epdqCaptureHandler.capture(buildTestCaptureRequest());
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProviderTest.java]: ditto previous comment

File: src/test/java/uk/gov/pay/connector/gateway/sandbox/SandboxPaymentProviderTest.java [hunk 5]
```diff
@@ -162,7 +164,7 @@ public void authorise_shouldGetGatewayErrorWhenCardNumDoesNotExistForAuthorisati
     @Test
     public void capture_shouldSucceedWhenCapturingAnyCharge() {
 
-        GatewayResponse<BaseCaptureResponse> gatewayResponse = provider.capture(CaptureGatewayRequest.valueOf(ChargeEntityFixture.aValidChargeEntity().build()));
+        GatewayResponse<BaseCaptureResponse> gatewayResponse = sandboxCaptureHandler.capture(CaptureGatewayRequest.valueOf(ChargeEntityFixture.aValidChargeEntity().build()));
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/gateway/sandbox/SandboxPaymentProviderTest.java]: ditto previous comment

File: src/test/java/uk/gov/pay/connector/gateway/smartpay/SmartpayPaymentProviderTest.java [hunk 6]
```diff
@@ -187,7 +189,7 @@ public void shouldCaptureAPaymentSuccessfully() {
                 .withGatewayAccountEntity(aServiceAccount())
                 .build();
 
-        GatewayResponse<WorldpayCaptureResponse> response = provider.capture(CaptureGatewayRequest.valueOf(chargeEntity));
+        GatewayResponse<SmartpayCaptureResponse> response = smartpayCaptureHandler.capture(CaptureGatewayRequest.valueOf(chargeEntity));
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/gateway/smartpay/SmartpayPaymentProviderTest.java]: ditto previous comment

File: src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java [hunk 7]
```diff
@@ -339,7 +342,7 @@ public void shouldSendSuccessfullyAnOrderForMerchant() throws Exception {
     public void shouldCaptureAPaymentSuccessfully() throws Exception {
         mockWorldpaySuccessfulCaptureResponse();
 
-        GatewayResponse response = provider.capture(getCaptureRequest());
+        GatewayResponse response = worldpayCaptureHandler.capture(getCaptureRequest());
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java]: ditto previous comment

File: src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java [hunk 8]
```diff
@@ -358,7 +361,7 @@ public void shouldErrorIfAuthorisationIsUnsuccessful() {
     @Test
     public void shouldErrorIfOrderReferenceNotKnownInCapture() {
         mockWorldpayErrorResponse(200);
-        GatewayResponse<WorldpayCaptureResponse> response = provider.capture(getCaptureRequest());
+        GatewayResponse<WorldpayCaptureResponse> response = worldpayCaptureHandler.capture(getCaptureRequest());
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java]: ditto previous comment

File: src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java [hunk 9]
```diff
@@ -368,7 +371,7 @@ public void shouldErrorIfOrderReferenceNotKnownInCapture() {
     @Test
     public void shouldErrorIfWorldpayResponseIsNot200() {
         mockWorldpayErrorResponse(400);
-        GatewayResponse<WorldpayCaptureResponse> response = provider.capture(getCaptureRequest());
+        GatewayResponse<WorldpayCaptureResponse> response = worldpayCaptureHandler.capture(getCaptureRequest());
```
Comment (oswaldquek) [inline src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java]: ditto previous comment

**Review by oswaldquek — CHANGES_REQUESTED** [review]

**Comment by oswaldquek** [comment]
Re the PR description, there is an explanation of "WHAT" and "HOW DID WE TEST THE CHANGES" but I think it'll be good to say "WHY" as well. I'm not really clear how this helps Stripe integration in the future as it seems a `GatewayResponseGenerator` class has been added but it's methods are used payment provider specific, e.g. getSmartpayGatewayResponse, getEpdqGatewayResponse, getWorldpayGatewayResponse.

File: src/main/java/uk/gov/pay/connector/gateway/CaptureHandler.java (hunk 1 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gateway/CaptureHandler.java]: 👍

File: src/test/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProviderTest.java (hunk 2 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProviderTest.java]: 👍 Good point. Moved to EpdqCaptureHandlerTest.

File: src/test/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProviderTest.java (hunk 3 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProviderTest.java]: Moved to EpdqCaptureHandlerTest.

File: src/test/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProviderTest.java (hunk 4 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/gateway/epdq/EpdqPaymentProviderTest.java]: Moved to EpdqCaptureHandlerTest.

File: src/test/java/uk/gov/pay/connector/gateway/sandbox/SandboxPaymentProviderTest.java (hunk 5 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/gateway/sandbox/SandboxPaymentProviderTest.java]: Moved capture tests to *CaptureHandlerTest

File: src/test/java/uk/gov/pay/connector/gateway/smartpay/SmartpayPaymentProviderTest.java (hunk 6 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/gateway/smartpay/SmartpayPaymentProviderTest.java]: Moved capture tests to *CaptureHandlerTest

File: src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java (hunk 7 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java]: Moved capture tests to *CaptureHandlerTest

File: src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java (hunk 8 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java]: Moved capture tests to *CaptureHandlerTest

File: src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java (hunk 9 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java]: Moved capture tests to *CaptureHandlerTest

**Review by oswaldquek — APPROVED** [review]

**Pull request merged**